### PR TITLE
adding herald rule and harbormaster for refs like tags and branches

### PR DESCRIPTION
**Pull request by goddenrich**
This revision introduces herald rules and harbormaster code to trigger builds on ref changes including tags and branches

### CODE REVIEW

**Comment by goddenrich** [comment]
This PR should allow us to set up herald rules to trigger harbormaster builds for ref events like pushing tags or branches. This allows us to run gitops such that changes to a repo trigger CI/CD builds

File: src/applications/diffusion/herald/HeraldPreCommitRefAdapter.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  public function getAdapterSortOrder() {
+    return 2000;
+  }
+
+  public function getAdapterContentDescription() {
+    return pht(
+      "React to branches and tags being pushed to hosted repositories.\n".
+      "Hook rules can block changes and send push summary mail.");
```
Comment (goddenrich) [inline src/applications/diffusion/herald/HeraldPreCommitRefAdapter.php]: This description should be changed

File: src/applications/diffusion/xaction/DiffusionPushLogBuildableTransaction.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
+<?php
+
+/* TM CHANGES */
+final class DiffusionCommitBuildableTransaction
```
Comment (goddenrich) [inline src/applications/diffusion/xaction/DiffusionPushLogBuildableTransaction.php]: wrong class! should be DiffusionPushLogBuildableTransaction

**Review by peterebden — COMMENTED** [review]
Are some of these files (e.g. PhabricatorRepositoryPushLog.php) getting copied in from upstream? If so we've historically put some `// TM changes` comments around them to help identify what's changed for future updates.

**Review by isobelormiston — APPROVED** [review]

**Comment by goddenrich** [comment]
> Are some of these files (e.g. PhabricatorRepositoryPushLog.php) getting copied in from upstream? If so we've historically put some `// TM changes` comments around them to help identify what's changed for future updates.

I think I have added that comment for everything changed

**Pull request merged**